### PR TITLE
feat: add pull request architecture review comments

### PR DESCRIPTION
**Pull request by NTCoding**
## Problem

Pull request reviewers must inspect the full code and test diff to identify changes to subdomains, entry points, use cases, aggregates, aggregate entities, and methods.

## Outcome

Each pull request receives one automatically maintained comment containing only architecture metadata that changed. A pull request with no tracked changes receives the minimal message `No architecture changes detected.`

## Changes

- Add a deterministic TypeScript source inspector and snapshot comparison.
- Group changes by subdomain, layer, Added or Removed, and aggregate ownership.
- Show roles and domain package locations without source links or generated interpretation.
- Add a `pull_request` workflow that executes an immutable reporter revision and parses base and head source archives as data.
- Create one marked comment and update that same comment after each new commit.
- Reject stale jobs before they can overwrite the latest comment.

## Acceptance criteria

- [x] The comment starts with the changed subdomains.
- [x] Empty layers and empty Added or Removed sections are omitted.
- [x] Aggregate entities and methods appear inside their aggregate.
- [x] Non aggregate domain changes appear in a table with name, role, and package.
- [x] A rename appears as the old name removed and the new name added.
- [x] Later commits update the existing comment instead of creating another comment.
- [x] Pull requests without tracked changes still receive the minimal marked comment.
- [x] No LLM or semantic summary is used.

## Validation

- `CI=true pnpm verify`
- `CI=true pnpm nx test docs`: 39 tests passed with 100% coverage
- [Action generated the full grouped happy path comment](https://github.com/NTCoding/living-architecture/actions/runs/33094808214) from a controlled commit containing multiple entry point, use case, aggregate, aggregate entity, method, and domain item changes.
- [Action updated the same comment after the controlled commit was reverted](https://github.com/NTCoding/living-architecture/actions/runs/33096111858).
- The maintained comment is [issue comment 5442211664](https://github.com/NTCoding/living-architecture/pull/477#issuecomment-5442211664).

## Design decisions and constraints

The workflow executes a reporter pinned to an immutable commit. It downloads the base and head source archives and parses them without executing pull request code. Workflow actions are pinned to immutable commits. A head SHA check prevents an older run from overwriting the latest result.

## Out of scope

Semantic move detection and generated review advice.

## Summary by CodeRabbit

* **New Features**
  * Added automated architecture reviews for pull requests.
  * Reviews compare entry points, use cases, and domain layers between the base and proposed changes.
  * Reports highlight added and removed architecture elements in Markdown, or confirm when no changes are detected.
  * Results appear in the pull request summary and are maintained in a bot comment.
  * Superseded review runs are automatically canceled to keep results current.

* **Tests**
  * Added coverage for report generation, validation, formatting, ownership tracking, and architecture-analysis edge cases.

### CODE REVIEW

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  return typeof tag?.comment === 'string' ? tag.comment.trim() : undefined
+}
+
+function typeReferencesName(
+  typeNode: ts.TypeNode | undefined,
+  expectedName: string,
+  aliases: ReadonlyMap<string, string>,
+): boolean {
+  if (typeNode === undefined) return false
+  return nodeReferencesName(typeNode, expectedName, aliases)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts]: Fixed in 59160e66. Aggregate ownership is keyed by package kind and aggregate name, and entity matching is limited to the same package. A same named aggregate regression test covers both package kinds.

File: .github/workflows/pr-architecture-review.yml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  issues: write
+
+jobs:
+  update-comment:
+    runs-on: ubuntu-latest
+    steps:
+      # Run only trusted code from the base revision. The pull request checkout
+      # below is parsed as source data and is never executed.
+      - name: Check out trusted reporter
+        uses: actions/checkout@d23441a48e516b6c34aea4fa41551a30e30af803 # v6
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply .github/workflows/pr-architecture-review.yml:25]: Keeping immutable commit SHAs. Major version tags are mutable and would weaken the workflow supply chain boundary; the pinned commits are intentional.

File: apps/docs/scripts/build-pull-request-architecture-review.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    head: { type: 'string' },
+    output: { type: 'string' },
+  },
+})
+
+const base = requiredArgument(values.base, '--base')
+const head = requiredArgument(values.head, '--head')
+const output = requiredArgument(values.output, '--output')
+const changes = compareArchitecture(inspectArchitecture(base), inspectArchitecture(head))
+writeFileSync(output, renderArchitectureReview(changes), 'utf8')
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/build-pull-request-architecture-review.ts]: Fixed in 59160e66. The command flow is now directly tested for a successful report write plus omitted and blank required arguments.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+): readonly AnnotatedDeclaration[] {
+  return sourceFile.statements.flatMap((statement) => {
+    if (ts.isVariableStatement(statement)) {
+      return annotatedVariables(statement, sourceFile, packageKind)
+    }
+    if (!isNamedArchitectureDeclaration(statement)) return []
+    const role = roleAnnotation(statement)
+    if (role === undefined || !isExported(statement)) return []
+    return [{ declaration: statement, name: statement.name.text, packageKind, role, sourceFile }]
+  })
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts]: Fixed in 59160e66. The flatMap callbacks now declare consistent readonly array return types. ESLint and full verification pass.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+): readonly string[] {
+  if (items.length === 0) return []
+  const includePackage = layer === 'domain'
+  return [
+    includePackage ? '| Name | Role | Package |' : '| Name | Role |',
+    includePackage ? '| --- | --- | --- |' : '| --- | --- |',
+    ...items.map((item) =>
+      includePackage
+        ? `| \`${item.name}\` | \`${item.role}\` | \`${item.packageKind}\` |`
+        : `| \`${item.name}\` | \`${item.role}\` |`,
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts]: Fixed in 59160e66. Markdown table values now escape pipes, backticks, and line breaks, with a regression test.

**Comment by NTCoding** [comment]
@coderabbitai review

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+function aggregateEntityOwners(
+  aggregates: readonly AnnotatedDeclaration[],
+  entities: readonly AnnotatedDeclaration[],
+): ReadonlyMap<string, readonly ArchitectureItem[]> {
+  const owners = new Map<string, ArchitectureItem[]>()
+  for (const entity of entities) {
+    const matchingAggregates = aggregates.filter(
+      (aggregate) =>
+        aggregate.packageKind === entity.packageKind && aggregateOwnsEntity(aggregate, entity.name),
+    )
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts:165]: Fixed in commit 1e17aa70. Aggregate entity matching now receives declaration source identity and resolves relative module specifiers before matching. A regression test uses two same named entities from distinct modules and verifies that each aggregate owns only its imported entity. The full verification gate passes.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  ]
+}
+
+function renderTableCode(value: string): string {
+  const escaped = value
+    .replaceAll('\r', ' ')
+    .replaceAll('\n', ' ')
+    .replaceAll('|', '&#124;')
+    .replaceAll('`', '&#96;')
+  return `\`${escaped}\``
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts]: Fixed in commit 1e17aa70. Markdown table values now use a backtick delimiter longer than the longest run, escape pipes with a backslash, and preserve consecutive backticks. The regression test covers a pipe, consecutive backticks, and a newline. The full verification gate passes.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    if (!ts.isImportDeclaration(statement) || !ts.isStringLiteral(statement.moduleSpecifier)) {
+      continue
+    }
+    const bindings = statement.importClause?.namedBindings
+    if (bindings === undefined || !ts.isNamedImports(bindings)) continue
+    for (const element of bindings.elements) {
+      imports.set(element.name.text, {
+        importedName: element.propertyName?.text ?? element.name.text,
+        moduleSpecifier: statement.moduleSpecifier.text,
+      })
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts]: Fixed in commit 7151edbc. Default imports are now recorded with default export identity and matched against the default modifier on the annotated declaration. The regression test imports a named default exported aggregate entity through a different local name and verifies that it is nested under the aggregate. The full verification gate passes.

**Comment by NTCoding** [comment]
@coderabbitai full review

**Comment by NTCoding** [comment]
@coderabbitai full review

File: .github/workflows/pr-architecture-review.yml [hunk 9]
```diff
@@ -0,0 +1,115 @@
+name: Pull request architecture review
+
+on:
+  pull_request:
+    types: [opened, reopened, synchronize]
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply .github/workflows/pr-architecture-review.yml:5]: Fixed in a1b22b14 and pinned in 6c1e69cd. Fork pull requests now use pull_request_target with the trusted reporter checkout, while fork source is downloaded only as data and never executed.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+      (subdomain) => `- [\`${subdomain.name}\`](#subdomain-${subdomain.name})`,
+    ),
+    '',
+    ...changes.subdomains.flatMap(renderSubdomain),
+  ].join('\n')
+}
+
+function renderSubdomain(subdomain: SubdomainArchitectureChanges): readonly string[] {
+  return [
+    `## Subdomain: \`${subdomain.name}\``,
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts]: Fixed in a1b22b14. Markdown values now use safe code span rendering, including backticks and line breaks, with a regression test for heading injection characters.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+  return [
+    `##### Aggregate: \`${aggregate.name}\``,
+    '',
+    ...renderAggregateMembers(
+      'Aggregate entities',
+      aggregate.entities.map((entity) => entity.name),
+    ),
+    ...renderAggregateMembers('Methods', aggregate.methods),
+    '',
+  ]
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts:94]: Fixed in a1b22b14. Aggregate headings now include the package kind, and a regression test covers the same aggregate name in domain model and published language.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+function entrypointRoots(appsRoot: string): readonly string[] {
+  return existsSync(appsRoot) ? findNamedDirectories(appsRoot, 'entrypoint') : []
+}
+
+function findNamedDirectories(directory: string, name: string): readonly string[] {
+  return readdirSync(directory, { withFileTypes: true }).flatMap((entry): readonly string[] => {
+    if (!entry.isDirectory()) return []
+    const child = path.join(directory, entry.name)
+    return entry.name === name ? [child] : findNamedDirectories(child, name)
+  })
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts:225]: Fixed in a1b22b14. Entrypoint discovery now shares the fixture directory exclusion, with regressions for fixture entrypoints and the executable script itself.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+export function publicMethodNames(declaration: AnnotatedDeclaration): readonly string[] {
+  if (!ts.isClassDeclaration(declaration.declaration)) return []
+  return uniqueText(
+    declaration.declaration.members
+      .filter(ts.isMethodDeclaration)
+      .filter((method) => !hasModifier(method, ts.SyntaxKind.PrivateKeyword))
+      .filter((method) => !hasModifier(method, ts.SyntaxKind.ProtectedKeyword))
+      .map((method) => method.name.getText(declaration.sourceFile)),
+  )
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts:78]: Fixed in a1b22b14. ECMAScript private methods are excluded, with a regression test for a public method beside a private method.

**Comment by NTCoding** [comment]
@coderabbitai full review

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+function inspectDomain(declarations: readonly AnnotatedDeclaration[]): ArchitectureLayerSnapshot {
+  const aggregateDeclarations = declarations.filter((entry) => entry.role === 'aggregate')
+  const aggregateEntities = declarations.filter((entry) => entry.role === 'aggregate-entity')
+  const entitiesByAggregate = aggregateEntityOwners(aggregateDeclarations, aggregateEntities)
+  const ownedEntities = new Set(
+    [...entitiesByAggregate.values()].flatMap((entities) => entities.map(itemKey)),
+  )
+  const aggregates = aggregateDeclarations.map((entry) => ({
+    entities: entitiesByAggregate.get(aggregateDeclarationKey(entry)) ?? [],
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts]: Fixed in 1e829a57 and pinned in 531d9f50. Ownership lookup and owned entity filtering now retain source declaration identity until the output boundary. The regression covers same named aggregates with different owned entities.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+      })
+    }
+    const bindings = importClause.namedBindings
+    if (bindings === undefined || !ts.isNamedImports(bindings)) continue
+    for (const element of bindings.elements) {
+      imports.set(element.name.text, {
+        importedName: element.propertyName?.text ?? element.name.text,
+        moduleSpecifier: statement.moduleSpecifier.text,
+      })
+    }
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts:270]: Fixed in 1e829a57 and pinned in 531d9f50. Namespace import bindings and qualified type references now resolve aggregate entity ownership. The regression covers a namespace imported entity type.

**Comment by NTCoding** [comment]
@coderabbitai full review

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+  }
+}
+
+function inspectPackageSubdomains(workspaceRoot: string): Map<string, MutableSubdomainSnapshot> {
+  const packagesRoot = path.join(workspaceRoot, 'packages')
+  const subdomains = new Map<string, MutableSubdomainSnapshot>()
+  if (!existsSync(packagesRoot)) return subdomains
+  for (const entry of readdirSync(packagesRoot, { withFileTypes: true })) {
+    if (!entry.isDirectory()) continue
+    const subdomainRoot = path.join(packagesRoot, entry.name)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-source.ts:49]: Fixed in 1520aacd and pinned in 62d56f3f. Package root discovery now excludes both fixture directory names, with a regression covering each name.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+    : modulePath
+  return [
+    modulePath,
+    `${extensionlessPath}.ts`,
+    `${extensionlessPath}.tsx`,
+    `${extensionlessPath}.mts`,
+    `${extensionlessPath}.cts`,
+    path.join(extensionlessPath, 'index.ts'),
+    path.join(extensionlessPath, 'index.tsx'),
+  ]
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts:307]: Fixed in 1520aacd and pinned in 62d56f3f. Production discovery now includes mts and cts sources, declaration exclusions cover their declaration file forms, and module candidates include their index forms. The regression covers both extensions.

**Comment by NTCoding** [comment]
@coderabbitai full review

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-source-edge-cases.spec.ts [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+  const workspace = mkdtempSync(path.join(tmpdir(), 'architecture-review-'))
+  temporaryWorkspaces.push(workspace)
+  return workspace
+}
+
+function writeWorkspaceFile(workspace: string, relativePath: string, source: string): void {
+  const filePath = path.join(workspace, relativePath)
+  mkdirSync(path.dirname(filePath), { recursive: true })
+  writeFileSync(filePath, source)
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-source-edge-cases.spec.ts]: Fixed in 41e9d853 and pinned in ca250e0d. Both source inspection suites now use one shared temporary workspace helper and cleanup implementation.

File: apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+    hasModifier(parameter, ts.SyntaxKind.PrivateKeyword) ||
+    hasModifier(parameter, ts.SyntaxKind.ProtectedKeyword) ||
+    hasModifier(parameter, ts.SyntaxKind.PublicKeyword) ||
+    hasModifier(parameter, ts.SyntaxKind.ReadonlyKeyword)
+  )
+}
+
+function hasModifier(node: ts.Node, modifier: ts.SyntaxKind): boolean {
+  return ts.canHaveModifiers(node) && ts.getModifiers(node)?.some((item) => item.kind === modifier)
+}
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/typescript-architecture-source.ts:347]: Fixed in 41e9d853 and pinned in ca250e0d. Modifier lookup now returns false explicitly when TypeScript provides no modifier list.

**Comment by NTCoding** [comment]
@coderabbitai full review

File: apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+function renderTableCode(value: string): string {
+  return renderCodeSpan(value, true)
+}
+
+function renderCodeSpan(value: string, escapePipes = false): string {
+  const normalized = value.replaceAll('\r', ' ').replaceAll('\n', ' ')
+  const singleLine = escapePipes ? normalized.replaceAll('|', '\\|') : normalized
+  const delimiter = '`'.repeat(longestBacktickRun(singleLine) + 1)
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review-markdown.ts:133]: Fixed in b4751509 and pinned in 71be249d. Table rendering now doubles existing backslashes before adding the pipe escape. The regression covers one and two backslashes before a pipe.

File: apps/docs/scripts/pull-request-architecture-review/architecture-review.spec.ts [hunk 21]
```diff
@@ -0,0 +1,435 @@
...
+import { inspectArchitecture } from './architecture-review-source'
+import { ArchitectureReviewSourceError } from './architecture-review-types'
+
+const temporaryWorkspaces: string[] = []
+
+afterEach(() => {
+  for (const workspace of temporaryWorkspaces.splice(0)) {
+    rmSync(workspace, { force: true, recursive: true })
+  }
+})
```
[the remark this answers is not in the document]
Comment (NTCoding) [reply apps/docs/scripts/pull-request-architecture-review/architecture-review.spec.ts]: Fixed in b4751509 and pinned in 71be249d. The main architecture review suite now uses the shared temporary workspace and cleanup helpers.

**Comment by NTCoding** [comment]
@coderabbitai full review

**Comment by NTCoding** [comment]
@coderabbitai full review
